### PR TITLE
WIP: Listen for broadcasts

### PR DESCRIPTION
**Pull request by mueller-ma**
This PR gives the user the ability to set up a custom broadcast
receiver that changes an item on receive.

This was designed to send Mi Band 2 button presses to the openHAB
server. This will only work if an app is installed, that send broadcasts
on button presses. Since Gadgetbridge does so, it is the default
broadcast.

I had to create a foreground service, that starts the receiver.
Otherwise it would be killed by the os. Foreground services have to show
a notification, but it shown as low priority, so it shouldnt disturb the
user.

Signed-off-by: mueller-ma <mueller-ma@users.noreply.github.com>

Reopen #388 
Related #257 

@kubawolanin I dont have tasker, so can you try if it is possible to send android broadcasts? If so, you can change one item.
[mobile-foss-debug.zip](https://github.com/openhab/openhab.android/files/1507074/mobile-foss-debug.zip)

Todo:
- [ ] Test if it works with tasker
- [ ] Start after boot
- [ ] Add better scalled icon for notification
- [ ] Add dropdown to select item
- [x] Change strings to let it sound more generall and less mi band

### CODE REVIEW

**Comment by kubawolanin** [comment]
Hey @mueller-ma!

Thank you so much for making it! I believe it will be a killer feature.

I couldn't figure out how to send broadcast from Tasker, though. I found some docs on [Intents](http://tasker.dinglisch.net/userguide/en/intents.html). Is the Android broadcast the same thing?
Sorry, my knowledge is limited here.

Here's some feedback on the overall experience with the feature:

1. It would be nice to have a list of all items (from `/rest/items`) to choose from. Typing it by hand is error prone and it would add a lot to the good UX :-)
1. Do you plan to extend the functionality to handle multiple broadcasts? It would be great, but then it would also require a separate view to store them.

Cheers

**Comment by ThomDietrich** [comment]
Hey guys!

Just wanted to agree, great initiative @mueller-ma !!
I don't know about the Mi band button but a Tasker integration would be amazing! Sadly I can not bring any experience with tasker-related android development to the table.

**Comment by rkoshak** [comment]
> I couldn't figure out how to send broadcast from Tasker, though. I found some docs on Intents. Is the Android broadcast the same thing?

I am about 90% certain that an Intent is what gets broadcast so to the Broadcast Listener. So if we have the proper format to form the Intent we should be able to broadcast to openHAB app. 

For example, to tell PocketCasts to start playing I send an intent with the Action

`au.com.shiftjelly.pocketcasts.action.PLAY`

and it is directed towards the Broadcast Receiver.

Similarly, to start Waze on my phone to automatically navigate to home I send an Intent with action

`android.intent.action.VIEW`

and Data

`waze://?ll=12.3456,-123.4567&navigate=yes`

and the Target is the Activity.

So we will need documentation for how to form the Intent to send to this receiver.

**Comment by ThomDietrich** [comment]
@rkoshak or you add a [Plugin](http://tasker.wikidot.com/plug-ins-and-3rd-party) to tasker, teaching it specific interactions your app provides. I believe that is the goal of this Issue? http://tasker.dinglisch.net/developers.html

**Comment by kubawolanin** [comment]
That's correct @ThomDietrich. 
@mueller-ma if you don't have Tasker let me show you real quick what the fuzz is all about :)

First you define some Tasks - they contain a sequence of Actions.
Tasks are then executed with Profiles (e.g. if you received a text message from certain number or something connected to phones Bluetooth etc)
screenshot_2017-11-29-19-55-41-901_net dinglisch android taskerm

Actions are the events (Intents?) like turning on the torch, airplane mode, launching an app etc. 
openHAB would be stored in 'Plugins' category.
screenshot_2017-11-29-19-56-01-673_net dinglisch android taskerm

If a certain app has a Tasker capability, it shows up in the Plugins category:
screenshot_2017-11-29-19-56-07-677_net dinglisch android taskerm

As you can see, I have the "Locale Toast Plug-In" from [this example code](http://www.twofortyfouram.com/developer/toast.zip) visible as well.

After you select a Plugin you'd like to work with, you'll be directed to plugin's configuration view which I roughly mocked-up in https://github.com/openhab/openhab.android/issues/257#issuecomment-269132020

Hope that this clears things up a bit :-)
Sorry if I'm off topic!

**Comment by rkoshak** [comment]
A Tasker plugin would indeed be a major improvement. Tasker is like openHAB for your phone and I know I and others go to great lengths to integrate the two together.

However, I think this and #257 are slightly different. Having a broadcast listener is attractive because it will work with other automation programs that can send intents as well. But I don't see how it buys us much in Tasker. Intents are considerably more arcane than using the REST API with an HTTP task so I'm thinking that a plugin might be the way to go for Tasker.

So in short, I think we need both this to support other automation tools on Android and #257 to make Tasker integration easier than it currently is.

That's my two cents, spoken as someone who doesn't have the time right now to actually contribute to either, soon hopefully though, once I work off the rest of my backlog of todos.

**Comment by mueller-ma** [comment]
Hey all,
Thanks for your feedback. I added some suggestions to the todo list.
I know tasker and I also know that is closed source, so dont expect too much support from me. I made this change very generall (you can configure broadcast and intent extra) so it may be used with gadgetbridge, tasker or any other app sending broadcasts. You could e.g. toggle a switch when battery runs low, you take a picture, unlock phone, etc.

FOSS tasker alternatives exists (https://github.com/biotinker/LibreTasks/issues), but development seems to be stuck.

Would be great to support multiple broadcasts, but I would need some help for that.

**Comment by mueller-ma** [comment]
I cannot figure out, why it doesnt start after boot.

**Comment by maniac103** [comment]
> I cannot figure out, why it doesnt start after boot.

You need to add the boot completed intent filter to a broadcast receiver, not the activity.

**Comment by mueller-ma** [comment]
@maniac103 Now the mainactivity starts on boot. I think I have to move that setup function to a seperate activity?

**Comment by maniac103** [comment]
Why don't you just start the service in the boot completed receiver?

**Comment by maniac103** [comment]
You mean duplication between startup on boot and handling of a changed pref? You could solve that by sending a broadcast to the boot completed receiver whenever the pref changes, so it can reevaluate the service state.

(BTW, I'd unify CustomBroadcastReceiver and BootCompletedReceiver into a single one that handles all aspects of that custom broadcast handling. Might be a personal preference thing only though.)

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java [hunk 1]
```diff
@@ -166,9 +167,9 @@ public void onFailure(Call call, int statusCode, Headers headers, byte[] respons
 
     // Loopj
 //    private static MyAsyncHttpClient mAsyncHttpClient;
-    private static MyAsyncHttpClient mAsyncHttpClient;
+    public static MyAsyncHttpClient mAsyncHttpClient;
     // Base URL of current openHAB connection
-    private String openHABBaseUrl = "http://demo.openhab.org:8080/";
+    public static String openHABBaseUrl = "http://demo.openhab.org:8080/";
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: Is making everything static here really a good idea? If it was me to decide, mAsyncHttpClient shouldn't be static to begin with, let alone public static. I'd suggest giving the service its own HTTP client and OpenHABTracker instance (like OpenHABVoiceService already has), and making CustomBroadcastReceiver a private non-static member class of that service so it has access to the client.

File: mobile/src/main/java/org/openhab/habdroid/util/Constants.java [hunk 2]
```diff
@@ -40,4 +40,18 @@
...
     }
-}
\ No newline at end of file
+    public static final String PREFERENCE_CUSTOM_BROADCAST    = "default_openhab_custom_broadcast_enabled";
+    public static final String PREFERENCE_CUSTOM_BROADCAST_BROADCAST = "default_openhab_custom_broadcast";
+    public static final String PREFERENCE_CUSTOM_BROADCAST_EXTRA = "default_openhab_custom_broadcast_extra";
+    public static final String PREFERENCE_CUSTOM_BROADCAST_ITEM = "default_openhab_custom_broadcast_item";
+
+    public interface ACTION {
+        public static String MAIN_ACTION = "org.openhab.habdroid.action.main";
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/util/Constants.java]: This doesn't seem to be actually used. It's set in the notification pending intent, but is never checked against.

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java [hunk 3]
```diff
@@ -338,10 +341,29 @@ private void processIntent(Intent intent) {
...
+            Log.d(TAG, "start cbr");
+            Intent startIntent = new Intent(this, ForegroundService.class);
+            startIntent.setAction(Constants.ACTION.STARTFOREGROUND_ACTION);
+            startIntent.putExtra("broadcast", mSettings.getString(Constants.PREFERENCE_CUSTOM_BROADCAST_BROADCAST, ""));
+            startService(startIntent);
+        } else {
+            Log.d(TAG, "stop cbr");
+            Intent stopIntent = new Intent(this, ForegroundService.class);
+            stopIntent.setAction(Constants.ACTION.STOPFOREGROUND_ACTION);
+            startService(stopIntent);
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: Just calling stopService() here (setting action not needed) should be sufficient. It's equivalent to calling stopSelf() in the service.

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java [hunk 4]
```diff
@@ -338,10 +341,29 @@ private void processIntent(Intent intent) {
...
             }
         }
     }
 
+    private void setupCustomBroadcastReceiver() {
+        // Custom broadcast receiver
+        if (mSettings.getBoolean(Constants.PREFERENCE_CUSTOM_BROADCAST, false)) {
+            Log.d(TAG, "start cbr");
+            Intent startIntent = new Intent(this, ForegroundService.class);
+            startIntent.setAction(Constants.ACTION.STARTFOREGROUND_ACTION);
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: Setting action isn't needed iMHO, the service can call startForeground() unconditionally in its onStartCommand().

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java [hunk 5]
```diff
@@ -338,10 +341,29 @@ private void processIntent(Intent intent) {
...
+            startIntent.putExtra("broadcast", mSettings.getString(Constants.PREFERENCE_CUSTOM_BROADCAST_BROADCAST, ""));
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: Can't the service figure out the broadcast it should listen to by itself?

File: mobile/src/main/java/org/openhab/habdroid/core/ForegroundService.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                    .setSmallIcon(R.drawable.icon_blank)
+                    .setPriority(Notification.PRIORITY_LOW)
+                    .setColor(ResourcesCompat.getColor(getResources(), R.color.openhab_orange, null))
+                    .setContentIntent(pendingIntent)
+                    .setOngoing(true).build();
+
+            startForeground(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE, notification);
+
+        } else if (intent.getAction().equals(Constants.ACTION.STOPFOREGROUND_ACTION)) {
+            // No API call available to check if a broadcast receiver is running: https://stackoverflow.com/questions/2682043/how-to-check-if-receiver-is-registered-in-android/3568906#3568906
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/ForegroundService.java]: You can track a boolean whether it's registered though. Seems cleaner than catching the exception.

File: mobile/src/main/java/org/openhab/habdroid/core/ForegroundService.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import android.preference.PreferenceManager;
+import android.support.v4.app.NotificationCompat;
+import android.support.v4.content.res.ResourcesCompat;
+import android.util.Log;
+
+import org.openhab.habdroid.R;
+import org.openhab.habdroid.ui.OpenHABMainActivity;
+import org.openhab.habdroid.util.Constants;
+
+public class ForegroundService extends Service {
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/ForegroundService.java]: Wouldn't this be better named `CustomBroadcastListenerService`?

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java (hunk 5 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: My idea was to pass all information (broadcast, item to change, intent extra, openhab url) from mainactivity to foreground service and from there to CustomBroadcastListenerService. But I dont have an idea how I can get those infos in CustomBroadcastReceiver#onReceive()

File: mobile/src/main/java/org/openhab/habdroid/util/Constants.java (hunk 2 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/util/Constants.java]: Removed that

File: mobile/src/main/java/org/openhab/habdroid/core/ForegroundService.java (hunk 7 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/core/ForegroundService.java]: I renamed it.

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java (hunk 5 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: CustomBroadcastListenerService now reads all settings and writes them to CustomBroadcastReceiver.

**Comment by mueller-ma** [comment]
Waiting for #562

**Comment by mueller-ma** [comment]
Closing for now as listening for only one broadcast is a too specific use case
